src/core/thread/csl_tx_scheduler.cpp

```diff
@@ -0,0 +1,271 @@
+                                                               Child &       aChild)
+{
+    return Get<IndirectSender>().PrepareFrameForChild(aFrame, aContext, aChild);
+}
+
+inline void CslTxScheduler::Callbacks::HandleSentFrameToChild(const Mac::TxFrame &aFrame,
+                                                              const FrameContext &aContext,
+                                                              otError             aError,
+                                                              Child &             aChild)
+{
+    Get<IndirectSender>().HandleSentFrameToChild(aFrame, aContext, aError, aChild);
+}
+
+//---------------------------------------------------------
+//
+CslTxScheduler::CslTxScheduler(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mCslTxChild(nullptr)
+    , mBusy(false)
+    , mFrameContext()
+    , mCallbacks(aInstance)
+{
+}
+
+void CslTxScheduler::Update(void)
+{
+    if (!mBusy)
+    {
+        RescheduleCslTx();
+    }
+    else
+    {
+        // `Mac` has already started the CSL tx, so wait for tx done callback
+        // to call `RescheduleCslTx`
+        if ((mCslTxChild != nullptr) && (mCslTxChild->GetIndirectMessage() != mFrameContext.mMessage))
+        {
+            mCslTxChild                      = nullptr;
+            mFrameContext.mMessage           = nullptr;
+            mFrameContext.mMessageNextOffset = 0;
+        }
+    }
+}
+
+void CslTxScheduler::Clear(void)
+{
+    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateAnyExceptInvalid))
+    {
+        child.SetCslTxAttempts(0);
+        child.SetCslSynchronized(false);
+        child.SetCslChannel(0);
+        child.SetCslSyncTimeout(0);
+        child.SetCslPeriod(0);
+        child.SetCslPhase(0);
+        child.SetCslLastHeard(TimeMilli(0));
+    }
+
+    mFrameContext.mMessageNextOffset = 0;
+    mFrameContext.mMessage           = nullptr;
+    mCslTxChild                      = nullptr;
+    mBusy                            = false;
+}
+
+void CslTxScheduler::RescheduleCslTx(void)
+{
+    uint64_t radioNow     = otPlatRadioGetNow(&GetInstance());
+    uint32_t minDelayTime = TimeMicro::kMaxDuration;
+    Child *  bestChild    = nullptr;
+
+    for (ChildTable::Iterator iter(GetInstance(), Child::StateFilter::kInStateAnyExceptInvalid); !iter.IsDone(); iter++)
+    {
+        Child &  child = *iter.GetChild();
+        uint32_t delay;
+
+        if (!child.IsCslSynchronized() || child.GetIndirectMessageCount() == 0 ||
+            child.GetCslTxAttempts() > kMaxCslTriggeredTxAttempts)
+        {
+            continue;
+        }
+
+        delay = GetNextCslTransmissionDelay(child, radioNow);
+        if (delay < minDelayTime)
+        {
+            minDelayTime = delay;
+            bestChild    = &child;
+        }
+    }
+
+    if (bestChild != mCslTxChild)
+    {
+        if (bestChild != nullptr)
+        {
+            Get<Mac::Mac>().RequestCslFrameTransmission(minDelayTime / 1000UL);
+        }
```

Comment:
Is it possible that the `bestChild` remains as before, but the `minDelayTime` is different. (e.g., is it possible the the CSL timing of a child (phase or period) change?).

Reply:
I think so. I think we can remove the check of `bestChild == mCslTxChild`. Thoughts?